File: src/nodes.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

enum class NodeType { AND = 1, OR = 2};


class Node {
    private:
        NodeType type;
        string courseTitle;
        string courseId;
        vector<Node*> children;
        bool value;

    public:
        Node(NodeType type, string courseTitle, string courseId) {
            this->type = type;
            this->courseTitle = courseTitle;
            this->courseId = courseId;
            this->value = false;
        }

        void addChild(Node* child) {
            if (child->getType() == NodeType::AND) {
                this->children.insert(this->children.begin(), child);
                return;
            }
            this->children.push_back(child);
        }

        void setValue(bool value) {
            this->value = value;
        }

        bool getValue() {
            return value;
        }

        NodeType getType() {
            return type;
        }

        string getCourseTitle() {
            return courseTitle;
        }

        string getCourseId() {
            return courseId;
        }

        vector<Node*> getChildren() {
            return children;
        }

};


vector<Node*> find_prereqs(vector<Node*> processed, vector<Node*> toBeProcessed) {
    // cout << "Processed: ";
    // for (Node* node : processed) {
    //     cout << node->getCourseTitle() << " ";
    // }
    // cout << endl;

    // cout << "To be processed: ";
    // for (Node* node : toBeProcessed) {
    //     cout << node->getCourseTitle() << " ";
    // }
    // cout << endl;

    if (toBeProcessed.size() == 0) {
        return processed;
    }

    Node* current = toBeProcessed[0];
    toBeProcessed.erase(toBeProcessed.begin());

    if (std::find(processed.begin(), processed.end(), current) == processed.end()) {
        processed.push_back(current);
    }

    if (current->getValue()) {
        return find_prereqs(processed, toBeProcessed);
    }

    if (current->getType() == NodeType::AND) {
        for (Node* child : current->getChildren()) {
            if (child->getType() == NodeType::AND) {
                toBeProcessed.insert(toBeProcessed.begin(), child);
            } else {
                toBeProcessed.push_back(child);
            }
        }
        return find_prereqs(processed, toBeProcessed);
    } 
    
    else {
        vector<Node*> mins = {};
        for (Node* child : current->getChildren()) {
            vector<Node*> temp_toBeProcessed = toBeProcessed;
            vector<Node*> temp_processed = processed;
            if (child->getType() == NodeType::AND) {
                temp_toBeProcessed.insert(temp_toBeProcessed.begin(), child);
            } else {
                temp_toBeProcessed.push_back(child);
            }
            vector<Node*> temp_mins = find_prereqs(temp_processed, temp_toBeProcessed);
            if (temp_mins.size() < mins.size() || mins.size() == 0) {
                mins = temp_mins;
            }
        }
        return mins;
    }

    return {new Node(NodeType::AND, "error", "error")};
}

void print_node(Node* node, int level) {
    for (int i = 0; i < level; i++) {
        cout << "  ";
    }
    cout << node->getCourseTitle() << " " << node->getValue() << endl;
    for (Node* child : node->getChildren()) {
        print_node(child, level + 1);
    }
}

File: api/models/nodes.go
package models

type NodeType bool

const (
	AND NodeType = true
	OR  NodeType = false
)

type Node struct {
	node_type    NodeType
	course_title string
	course_id    string
	children     []*Node
	value        bool
}

func NewNode(node_type NodeType, course_title string, course_id string) *Node {
	return &Node{
		node_type:    node_type,
		course_title: course_title,
		course_id:    course_id,
		children:     []*Node{},
		value:        false,
	}
}

func (n *Node) AddChild(child *Node) {
	if child.node_type == AND {
		n.children = append([]*Node{child}, n.children...)
	} else {
		n.children = append(n.children, child)
	}
}

func (n *Node) SetValue(value bool) {
	n.value = value
}

func (n *Node) GetValue() bool {
	return n.value
}

func (n *Node) GetType() NodeType {
	return n.node_type
}

func (n *Node) GetCourseTitle() string {
	return n.course_title
}

func (n *Node) GetCourseId() string {
	return n.course_id
}

func (n *Node) GetChildren() []*Node {
	return n.children
}

func FindPrereqs(processed []*Node, toBeProcessed []*Node) []*Node {
	if len(toBeProcessed) == 0 {
		return processed
	}

	current := toBeProcessed[0]
	toBeProcessed = toBeProcessed[1:]

	// If the current node is already satisfied, skip further processing
	if current.GetValue() {
		return FindPrereqs(processed, toBeProcessed)
	}

	// Check if already processed
	if !Contains(processed, current) {
		processed = append(processed, current)
	}

	if current.GetType() == AND {

		for _, child := range current.GetChildren() {
			if child.GetType() == AND {
				toBeProcessed = append([]*Node{child}, toBeProcessed...)
			} else {
				toBeProcessed = append(toBeProcessed, child)
			}
		}
		return FindPrereqs(processed, toBeProcessed)
	} else {
		var mins []*Node
		for _, child := range current.GetChildren() {
			tempToBeProcessed := append([]*Node{}, toBeProcessed...)
			tempProcessed := append([]*Node{}, processed...)

			if child.GetType() == AND {
				tempToBeProcessed = append([]*Node{child}, tempToBeProcessed...)
			} else {
				tempToBeProcessed = append(tempToBeProcessed, child)
			}

			tempMins := FindPrereqs(tempProcessed, tempToBeProcessed)
			if len(tempMins) < len(mins) || len(mins) == 0 {
				mins = tempMins
			}
		}
		return mins
	}
}

func Contains(nodes []*Node, target *Node) bool {
	for _, n := range nodes {
		if n == target {
			return true
		}
	}
	return false
}

func PrintNode(n *Node, level int) {
	for i := 0; i < level; i++ {
		print("  ")
	}
	println(n.GetCourseTitle(), n.GetValue())
	for _, child := range n.GetChildren() {
		PrintNode(child, level+1)
	}
}
